File: src/main/java/org/pyjjs/scheduler/core/common/locale/LocaleResolver.java
package org.pyjjs.scheduler.core.common.locale;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.util.Properties;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

public class LocaleResolver {
    private static final Logger LOG = LoggerFactory.getLogger(LocaleResolver.class);
    private static ConcurrentMap<String, LocaleResolver> instance = new ConcurrentHashMap<>();

    private Properties properties;

    private LocaleResolver(LangResolver language) throws IOException {
        properties = new Properties();
        String propFileName = language.getFileName();
        properties.load(getClass().getClassLoader().getResourceAsStream(propFileName));
    }

    public Properties values() {
        return properties;
    }

    public String getString(String key, Object... params) {
        String rawString = properties.getProperty(key);
        try {
            return String.format(new String(rawString.getBytes("ISO-8859-1"), "UTF-8"), params);
        } catch (UnsupportedEncodingException e) {
            throw new AssertionError("Encoding is incorrect",e);
        }
    }

    public static LocaleResolver get(LangResolver langResolver) throws LocaleNotFoundException {
        LocaleResolver localeResolver = instance.get(langResolver.getShortLang());
        if(localeResolver == null) {
            synchronized (LocaleResolver.class) {
                localeResolver = instance.get(langResolver.getShortLang());
                if(localeResolver == null) {
                    try {
                        localeResolver = new LocaleResolver(langResolver);
                    } catch (IOException e) {
                        throw new LocaleNotFoundException(e.getMessage());
                    }
                    instance.put(langResolver.getShortLang(), localeResolver);
                }
            }
        }
        return localeResolver;
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/common/locale/LocaleResolver.kt
package org.pyjjs.scheduler.core.common.locale

import java.io.IOException
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class LocaleResolver private constructor(language: LangResolver) {

    private val properties: Properties

    init {
        properties = Properties()
        val propFileName = language.fileName
        properties.load(javaClass.classLoader.getResourceAsStream(propFileName))
    }

    fun values(): Properties {
        return properties
    }

    fun getString(key: String, vararg params: Any): String {
        val rawString: String = properties.getProperty(key)
        val byteArray = rawString.toByteArray(charset("ISO-8859-1"))
        return String.format(byteArray.toString(charset("UTF-8")), *params)
    }

    companion object {
        private val instance = ConcurrentHashMap<String, LocaleResolver>()

        @Throws(LocaleNotFoundException::class)
        operator fun get(langResolver: LangResolver): LocaleResolver {
            var localeResolver: LocaleResolver? = instance[langResolver.shortLang]
            if (localeResolver == null) {
                synchronized(LocaleResolver::class.java) {
                    localeResolver = instance[langResolver.shortLang]
                    if (localeResolver == null) {
                        try {
                            localeResolver = LocaleResolver(langResolver)
                        } catch (e: IOException) {
                            throw LocaleNotFoundException(e.message ?: "Locale not found")
                        }

                        instance.put(langResolver.shortLang, localeResolver as LocaleResolver)
                    }
                }
            }
            return localeResolver as LocaleResolver
        }
    }
}
